### PR TITLE
integrate: add simpson's rule

### PR DESCRIPTION
**Pull request by nisfornature**
This pull request (carried on from #1100) is a Simpson's method implementation for #400.

### CODE REVIEW

File: integrate/simpsons.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+	case len(x) <= 2:
+		panic("integrate: input data too small")
+	case !sort.Float64sAreSorted(x):
+		panic("integrate: must be sorted")
+	}
+
+	var integral float64
+
+	for i := 1; i < len(x)-1; i += 2 {
+
```
Comment (kortschak) [inline integrate/simpsons.go]: Please delete this blank line.

File: integrate/simpsons.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+//             6 * h_{N-2} * (h_{N-2} + h_{N-1})
+//  {a_1}_i = h^{2}_{N-1} + 3 * h^{3}_{N-1} * h_{N-2} /
+//            6 * h_{N-2}
+//  {a_2}_i = 2 * h^{2}_{N-1} + 3 * h^{2}_{N-1} * h_{N-2} /
+//            6 * (h_{N-2} + h_{N-1})
+// More information is available at:
+// https://en.wikipedia.org/wiki/Simpson%27s_rule#Composite_Simpson's_rule_for_irregularly_spaced_data
+//
+// The (x,f) input data points must be sorted along x. One can use stat.SortWeighted to do that.
+// The x and f slices must be of equal length and have length > 1.
```
Comment (kortschak) [inline integrate/simpsons.go]: s/> 1/greater than one/

Comment (vladimir-ch) [reply integrate/simpsons.go]: greater than two

Comment (kortschak) [reply integrate/simpsons.go]: Indeed.

Comment (nisfornature) [reply integrate/simpsons.go]: Oops. I made the following changes? What additional changes are needed?

- Updated length
- Reduced spacing
- Defined a,b
- Defined h_{k}
- matched the description of the first formula to the implementation `\sum_{i=1}^{N} {a_0}_i * f_{2i} + {a_1}_i * f_{i} + {a_2}_i * f_{i+1}`

```
// Simpson's method approximates the integral of a function f(x)
// by means of subdividing the interval of integration into segments
// and applying a method that fits a polynomial to each subinterval.
// This implementation makes use of the following composite Simpson's method
// to estimate \int_a^b f(x) dx where a, b are x at x[0] and x[len(x)-1] respectively:
//  \sum_{i=1}^{N} {a_0}_i * f_{2i} + {a_1}_i * f_{i} + {a_2}_i * f_{i+1}
// where N is the count of subintervals and {a_0}_i, {a_1}_i, and {a_2}_i are constants at index i given by:
//  {a_0}_i * f_{N-2} + {a_1}_i * f_{N-1} + {a_2}_i * f_{N}
// where the coefficients are:
//  {a_0}_i = 2 * h^{3}_{i} - h^{3}_{i+1} + 3 * h_{i+1} * h^{2}_{i} /
//            6 * h_{i} * (h_{i} + h_{i+1})
//  {a_1}_i = h^{3}_{i} + h^{3}_{i+1} +
//            3 * h_{i} * h_{i+1} * (h_{i} + h_{i+1}) /
//            6 * h_{i} * h_{i+1}
//  {a_2}_i = -h^{3}_{i} + 2 * h^{3}_{i+1} + 3 * h_{i} * h^{2}_{i+1} /
//            6 * h_{i+1} * (h_{i} + h_{i+1})
// where h_{k} is the difference x[k] - x[(k-1)].
// The formula above approximates the integral of function of f if N is an even number of subintervals.
// If the number of subintervals are odd, the subintervals up i=0..n-2 are given by the above
// and the approximations over the second to last and last subintervals are given by:
//  {a_0}_i * f_{N-2} + {a_1}_i * f_{N-1} + {a_2}_i * f_{N}
// where the coefficients are:
//  {a_0}_i = -1 * h^{3}_{N-1} /
//             6 * h_{N-2} * (h_{N-2} + h_{N-1})
//  {a_1}_i = h^{2}_{N-1} + 3 * h^{3}_{N-1} * h_{N-2} /
//            6 * h_{N-2}
//  {a_2}_i = 2 * h^{2}_{N-1} + 3 * h^{2}_{N-1} * h_{N-2} /
//            6 * (h_{N-2} + h_{N-1})
// More information is available at:
// https://en.wikipedia.org/wiki/Simpson%27s_rule#Composite_Simpson's_rule_for_irregularly_spaced_data
//
// The (x,f) input data points must be sorted along x. One can use stat.SortWeighted to do that.
// The x and f slices must be of equal length and the length must be greater than two.
```

File: integrate/simpsons.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
+// Copyright ©2019 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package integrate
+
+import (
```
Comment (kortschak) [inline integrate/simpsons.go]: `import "sort"`

File: integrate/simpsons_test.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+	"math"
+	"testing"
+
+	"gonum.org/v1/gonum/floats"
+)
+
+func TestSimpsons(t *testing.T) {
+	const (
+		n   = 1e6
+		tol = 1e-12
```
Comment (kortschak) [inline integrate/simpsons_test.go]: Please delete this and just use a float literal in the test cases.

File: integrate/simpsons_test.go [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+	)
+	x := floats.Span(make([]float64, n), 0, 1)
+	for i, test := range []struct {
+		x         []float64
+		f         func(x float64) float64
+		want      float64
+		tolerance float64
```
Comment (kortschak) [inline integrate/simpsons_test.go]: s/tolerance/tol/g

File: integrate/simpsons_test.go [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+
+	"gonum.org/v1/gonum/floats"
+)
+
+func TestSimpsons(t *testing.T) {
+	const (
+		n   = 1e6
+		tol = 1e-12
+	)
+	x := floats.Span(make([]float64, n), 0, 1)
```
Comment (kortschak) [inline integrate/simpsons_test.go]: Same here; a literal expression in the test case is better.

**Comment by btracey** [comment]
Thank you very much for the PR!

I think the merging is blocked waiting for @kortschak ?

Have you signed the AUTHORS and CONTRIBUTORS? If not, could you submit a PR doing so, and we'll merge these together?

**Comment by kortschak** [comment]
I'll merge these then the PR against A+C is ready.

**Comment by nisfornature** [comment]
@kortschak, add yourself twice? I'll be more diligent in the future when submitting PRs.

**Comment by kortschak** [comment]
Thanks

**Pull request merged**